File: rb101_rb109/medium2/triangle_sides.rb
=begin
input: 3 integers
output: a symbol
rules:
  - Takes the lengths of the 3 sides of a triangle as arguments, and returns a symbol :equilateral, :isosceles, :scalene, or :invalid 
  depending on whether the triangle is equilateral, isosceles, scalene, or invalid.
  - To be a valid triangle, the sum of the lengths of the two shortest sides must be greater than the length of the longest side, 
    and all sides must have lengths greater than 0
algorithm:
  - valid triangle:
    - create an array with 3 arguments as items
    - check if the 3 arguments are greater than 0
    - check the sorted arguments if the addition of frist 2 items are greater than the 3 item
  - type of triangle:
    - sort the 3 arguments :
      - return equilateram if array[0] == array[1] && array[1] == array[2]
      - return isosceles if array[0] == array[1]
      - return scalene for the rest
=end
# first attempt
def valid_triangle?(arr)
  arr.all? {|n| n > 0} && arr.sort[0..1].sum > arr.sort[2]
end

def triangle(num1, num2, num3)
  arr = [num1, num2, num3]
  arr = arr.sort
  if valid_triangle?(arr)
    if arr[0] == arr[1] && arr[1] == arr[2]
      return :equilateral
    elsif arr[0] == arr[1] || arr[1] == arr[2]
      return :isosceles
    else
      :scalene
    end
  else
    :invalid
  end
end

# another soultion: use case statement
=begin
def valid_triangle?(arr)
  arr.all? {|n| n > 0} && arr.sort[0..1].sum > arr.sort[2]
end

def triangle(num1, num2, num3)
  arr =[num1, num2, num3]
  arr = arr.sort
  case
  when !valid_triangle?(arr)
    :invalid
  when arr[0] == arr[1] && arr[1] == arr[2]
    :equilateral
  when arr[0] == arr[1] || arr[1] == arr[2]
    :isosceles
  else
    :scalene
  end
end
=end

p triangle(3,3,3)
p triangle(3,3,1.5)
p triangle(3,4,5)
p triangle(0,3,3)
p triangle(3,1,1)

File: ls215/medium2_2nd/triangle_sides.js
/*
18:23 -18: 44

input: 3 numbers (>= 0) 
output: one of the 4 string
rules:
  valid triangle: 
    - sum of the lengths of the two shortest sides > longest length
    - all sides > 0
  Equilateral: All three sides are of equal length.
  Isosceles: Two sides are of equal length, while the third is different.
  Scalene: All three sides are of different lengths.
data: array of lengths, sorted short to longest

algo:
  create array of all sides, sorted by length (short to long);
  check if valid triangle:
    - every side is > 0
    - arr[0] + arr[1] > arr[2]
  if arr[0] === arr[1] && arr[1] === arr[2] => equi
  if arr[0] === arr[1] || arr[1] === arr[2] => iso
  else => scal

*/

function triangle(side1, side2, side3) {
  let sides = [side1, side2, side3].sort((a,b) => a - b)

  if (!validTriangle(sides)) {
    return 'invalid';
  }

  if (sides[0] === sides[1] && sides[1] === sides[2]) {
    return 'equilateral';
  } else if (sides[0] === sides[1] || sides[1] === sides[2]) {
    return 'isosceles';
  } else {
    return 'scalene';
  }
  
}

function validTriangle(sides) {
  return sides.every(side => side > 0) && sides[0] + sides[1] > sides[2];
}

console.log(triangle(3, 3, 3));        // "equilateral"
console.log(triangle(3, 3, 1.5));      // "isosceles"
console.log(triangle(3, 4, 5));        // "scalene"
console.log(triangle(0, 3, 3));        // "invalid"
console.log(triangle(3, 1, 1));        // "invalid"